include/nccl_ofi_rdma.h

```diff
@@ -659,6 +679,10 @@ typedef struct nccl_net_ofi_rdma_device {
 
 	/* Memory registration key pool */
 	nccl_ofi_idpool_t key_pool;
+
+#if HAVE_NVTX_TRACING
```

Comment:
&& NCCL_OFI_NVTX_TRACE_PER_DEV ?

Reply:
Grr... this doesn't work, because I have defines written like this:
```
#define NCCL_OFI_TRACE_SEND_CTRL_RECV_NVTX(dev, rail_id, comm, msg_seq_num) do { \
        nvtxDomainHandle_t handle; \
        if (NCCL_OFI_NVTX_TRACE_PER_COMM) { \
                handle = ((nccl_net_ofi_rdma_send_comm_t*)comm)->nvtx_domain[msg_seq_num % N_NVTX_DOMAIN_PER_COMM]; \
                nvtx_mark_domain(handle, "Send_ctrl_recv", 0x00ffff); \
        } \
        if (NCCL_OFI_NVTX_TRACE_PER_DEV) { \
                handle = ((nccl_net_ofi_rdma_device_t*)(comm->base.base.ep->device))->nvtx_domain[rail_id]; \
                nvtx_mark_domain(handle, "Send_ctrl_recv", 0x00ffff); \
        } \
} while (0)
```

Which will error if `device->nvtx_domain` isn't defined. And putting an `#if` inside a `#define` is complicated.

Reply:
builtin_choose_expr can help do the right thing.

Reply:
Can you give a small code snippet? I don't understand how to use `builtin_choose_expr` here to express a multi-line statement.

Reply:
The __builtin_choose_expr is a GCC built-in function that allows you to choose between two expressions based on a condition at compile-time. It can be used to select a single statement or a block of statements (multiple lines) based on a constant expression.

Here's the syntax for using __builtin_choose_expr to choose a single-line statement:

`__builtin_choose_expr(constant_condition, true_expression, false_expression);`

To choose a multi-line statement or a block of code, you need to use a compound statement (enclosed in curly braces {}). Here's an example:

```
#include <stdio.h>
#define CONDITION 1

int main() {
    __builtin_choose_expr(CONDITION, ({
        printf("Condition is true\n");
        printf("This is a multi-line statement\n");
        42; // Return value of the compound statement
    }), ({
        printf("Condition is false\n");
        24; // Return value of the compound statement
    }));
    return 0;
}
```

In this example, if CONDITION is defined as 1 (true), the first compound statement ({ ... }) will be executed, printing "Condition is true" and "This is a multi-line statement". If CONDITION is defined as 0 (false), the second compound statement will be executed, printing "Condition is false".

The compound statements ({ ... }) are a GCC extension that allows you to group multiple statements together and treat them as a single expression. The last expression inside the compound statement is the value that the entire expression evaluates to (in this case, 42 or 24).

Reply:
Do not block on this.